torch/distributed/distributed_c10d.py

```diff
@@ -147,7 +147,7 @@ def __getattribute__(self, key):
 
 
 class group(object):
-    WORLD = object()
+    WORLD: Optional[ProcessGroup] = None
```

Comment:
Does this make the below `GroupMember` class sort of strange in that it is like an enum, but GroupMember.WORLD is a ProcessGroup and GroupMember.NON_GROUP_MEMBER is an object()?

Reply:
Nit: maybe add a comment saying that it will be the default pg once initialized?

Reply:
> Does this make the below GroupMember class sort of strange in that it is like an enum, but GroupMember.WORLD is a ProcessGroup and GroupMember.NON_GROUP_MEMBER is an object()?

Yes, it does although this is unfortunately due to how the code was initially written and its hard to change some of this due to backward compatibility reasons. Both classes `group` and `GroupMember` are public and are used by PyTorch users. `NON_GROUP_MEMBER` is particularly weird for the fact that we actually return that to `new_group` calls which are not part of the new_group. As a result, those calls don't get ProcessGroup returned (or None) but this arbitrarty object `NON_GROUP_MEMBER` that is used elsewhere to perform checks.